### PR TITLE
[Test Framework] Add initial support for Matchers

### PR DESCRIPTION
**Pull request by SupunS**
Work towards https://github.com/onflow/cadence/issues/331

## Description

Supports:
- **Equal matcher**
  ```cadence
  // matcher with primitive type
  let matcher = Test.equal(1)

  // matcher with struct type
  let matcher = Test.equal(Foo())
  ```

- **Custom matchers**: 
  Supports defining custom matchers that deal with `AnyStruct` type.

  _e.g.(1): Customer matcher for a primitive type_
  ```cadence
  let matcher = Test.NewMatcher(fun (_ value: Int): Bool {
      return value == 7
  })

  assert(matcher.test(7))
  ```
   _e.g.(2): Customer matcher for a struct type_
  ```cadence
  let matcher = Test.NewMatcher(fun (_ value: AnyStruct): Bool {
      if !value.isInstance(Type<Int>()) {
          return false
      }

      return (value as! Int) > 5
  })

  assert(matcher.test(8))
  ```
   _e.g.(3): Customer matcher for handling **resource type**. Even though resource types are not supported, it can be made to work with a resource reference_
  ```cadence
  let matcher = Test.NewMatcher(fun (_ value: &Foo): Bool {
      return value.a == 4
  })

  let f <-create Foo(4)
  assert(matcher.test(&f as &Foo))
  ```

- **Expect function**
  ```cadence
  let matcher = Test.equal("other string")
  Test.expect("this string", matcher)
  ```
  _*This API was slightly changed compared to the proposal, to support resources. This API is not final and can be adjusted based on requirements/feedback._

______

- [ ] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/interpreter/interpreter.go [hunk 1]
```diff
@@ -1085,10 +1085,15 @@ func (interpreter *Interpreter) InvokeExternally(
 		preparedArguments[i] = interpreter.ConvertAndBox(getLocationRange, argument, nil, parameterType)
 	}
 
+	var self *CompositeValue
+	if boundFunc, ok := functionValue.(BoundFunctionValue); ok {
+		self = boundFunc.Self
+	}
```
Comment (dsainati1) [inline runtime/interpreter/interpreter.go:1091]: Does this have any effect on anything else outside of the testing framework? Seems like changing the receiver of an externally invoked function could have other consequences.

File: runtime/interpreter/interpreter.go [hunk 2]
```diff
@@ -1085,10 +1085,15 @@ func (interpreter *Interpreter) InvokeExternally(
...
+
 	// NOTE: can't fill argument types, as they are unknown
 	invocation := NewInvocation(
 		interpreter,
-		nil,
+		self,
```
Comment (turbolent) [inline runtime/interpreter/interpreter.go:1096]: :+1:

File: runtime/stdlib/test.go [hunk 3]
```diff
@@ -281,6 +442,80 @@ var testNewEmulatorBlockchainFunction = interpreter.NewUnmeteredHostFunctionValu
...
 
+// 'Test.NewMatcher' function.
+// Constructs a matcher that test only 'AnyStruct'.
+// Accepts test function that accepts subtype of 'AnyStruct'.
+//
+// Signature:
+//    Test.NewMatcher<T>(test: ((T): Bool)): AnyStruct<Test.Matcher>
+// where T is optional, and bound to 'AnyStruct'.
+//
+// Sample usage: Test.NewMatcher(fun (_value: Int: Bool) { return true})
```
Comment (turbolent) [inline runtime/stdlib/test.go:453]:
```suggestion
// Signature:
//    fun Test.NewMatcher<T: AnyStruct>(test: ((T): Bool)): AnyStruct{Test.Matcher}
// where `T` is optional, and bound to `AnyStruct`.
//
// Sample usage: `Test.NewMatcher(fun (_ value: Int: Bool) { return true })`
```

File: runtime/stdlib/test.go [hunk 4]
```diff
@@ -1016,6 +1251,112 @@ var emulatorBackendCommitBlockFunction = interpreter.NewUnmeteredHostFunctionVal
...
+			func(invocation interpreter.Invocation) interpreter.Value {
+
+				thisValue, ok := invocation.Arguments[0].(interpreter.EquatableValue)
+				if !ok {
+					panic(errors.NewUnreachableError())
+				}
+
+				equal := thisValue.Equal(
+					inter,
+					interpreter.ReturnEmptyLocationRange,
```
Comment (turbolent) [inline runtime/stdlib/test.go:1346]:
```suggestion
					invocation.GetLocationRange,
```

File: runtime/stdlib/test.go [hunk 5]
```diff
@@ -1033,3 +1374,251 @@ func (e TestFailedError) Unwrap() error {
...
+// 'Test.DefaultMatcher' struct.
+//
+// This is the default implementation of 'Test.Matcher' interface.
+// It accepts 'Any' for the test function. Hence, can be used with both structs and resources.
+// But the usage is limited by the constructor.
+// i.e: the constructor is hidden from users, to avoid misuses (see below).
+// Instead, use 'Test.NewMatcher()' to construct a matcher that test only 'AnyStruct'.
+//
+// Limitations:
+// Handling resource inside matchers is not supported yet, only exception being the built-in matchers.
```
Comment (turbolent) [inline runtime/stdlib/test.go:1387]: Can `AnyStruct` be used instead then?

File: runtime/testlib_test.go [hunk 6]
```diff
@@ -114,3 +138,697 @@ func TestExecuteScript(t *testing.T) {
...
+		script := `
+            import Test
+
+            pub fun main() {
+
+                let matcher = Test.NewMatcher(fun (_ value: Int): Bool {
+                     return value == 7
+                })
+
+                assert(matcher.test(7))
```
Comment (turbolent) [inline runtime/testlib_test.go:187]: Just to clarify my understanding: `matcher` has non-generic type `AnyStruct{Test.Matcher}` here, correct?

If that's the case, then `matcher.test` could be called with a non-`Int` value here – is the run-time type checked somewhere? (I can't quite see it)

File: runtime/testlib_test.go [hunk 7]
```diff
@@ -114,3 +138,697 @@ func TestExecuteScript(t *testing.T) {
...
+
+            pub fun main() {
+
+                let matcher = Test.NewMatcher(fun (_ value: &Foo): Bool {
+                    return value.a == 4
+                })
+
+                let f <-create Foo(4)
+
+                assert(matcher.test(&f as &Foo))
```
Comment (turbolent) [inline runtime/testlib_test.go:215]: 😍

**Review by turbolent — COMMENTED** [review]
Great work!

File: runtime/interpreter/interpreter.go (hunk 1 above)
Comment (SupunS) [reply runtime/interpreter/interpreter.go:1091]: It shouldn't really affect others. `InvokeExternally` is used to invoke a cadence function from Go. For bound functions (e.g: `foo.bar()`), the receiver information was missing. Perhaps it was never needed before.

File: runtime/testlib_test.go (hunk 6 above)
Comment (SupunS) [reply runtime/testlib_test.go:187]: Aha, good catch! This is not handled at the moment. The test-case I had added for this is flawed (it checks equality, so it fails anyway). Need to add a runtime check.
Enforcing that statically would have been much better, but I don't see a way to do that. Do you have any ideas?

File: runtime/stdlib/test.go (hunk 5 above)
Comment (SupunS) [reply runtime/stdlib/test.go:1281]: Yeah we could. The only reason to add `Any` here was to allow some built-in matchers like `equal`, `beNil`, etc. to work with resources. If we are to say that we aren't going to support even those with resources (custom matchers are anyway not supported), then we can just make this `AnyStruct`

File: runtime/testlib_test.go (hunk 6 above)
Comment (SupunS) [reply runtime/testlib_test.go:187]: added the runtime check with 4792dd12f0565d7972d43f0fef7a50661afe69fb

File: runtime/stdlib/test.go (hunk 5 above)
Comment (turbolent) [reply runtime/stdlib/test.go:1281]: Maybe we can use references for resources?

File: runtime/testlib_test.go (hunk 6 above)
Comment (turbolent) [reply runtime/testlib_test.go:187]: Nice! Thank you for also adding a test

File: runtime/stdlib/test.go (hunk 5 above)
Comment (SupunS) [reply runtime/stdlib/test.go:1281]: The problem with references is that equality works differently. It checks whether the reference is equal (i.e: they refer to the same resource), but not whether the two resources referred to are equal.

Comment (SupunS) [reply runtime/stdlib/test.go:1281]: yea, that makes sense

Comment (SupunS) [reply runtime/stdlib/test.go:1446]: updated in e0d14e7026c9e533dcbf95fae917631ce580e9e7 to use `AnyStruct`

**Review by turbolent — APPROVED** [review]
Nice!

**Comment by SupunS** [comment]
@turbolent Thanks for the review! I did a quick refactor to remove the `Matcher` interface because it was added in the first place to support resources, and is no longer needed / adds complexity.

Can you have one last look? It's mostly this commit: 58f9b11908eefed4c47e71b89bd7bf529b5b620a

**Review by turbolent — APPROVED** [review]
Looks good!

**Pull request merged**